### PR TITLE
Make test runner support the latest TypeDB Cluster

### PR DESCRIPTION
**Pull request by lolski**
## What is the goal of this PR?

The Cluster test runner supports the latest version of TypeDB Cluster.

Additionally, the runner can be configured more flexibly, allowing for any server parameters to be overridden.

## What are the changes implemented in this PR?

1. Support the latest TypeDB Cluster by updating how the runner boots up the server.
2. Make configuring the runner more flexible by introducing additional factory method variants.

### CODE REVIEW

File: test/server/TypeDBClusterRunner.java [hunk 1]
```diff
@@ -18,34 +18,70 @@
 
 package com.vaticle.typedb.common.test.server;
 
-import com.vaticle.typedb.common.collection.Triple;
-
+import javax.sound.sampled.Port;
```
Comment (lolski) [inline test/server/TypeDBClusterRunner.java:21]: wot

File: test/server/TypeDBClusterRunner.java [hunk 2]
```diff
@@ -18,34 +18,70 @@
...
     }
 
-    public TypeDBClusterRunner() throws InterruptedException, TimeoutException, IOException {
-        this(ThreadLocalRandom.current().nextInt(40000, 60000));
+    public static TypeDBClusterRunner create(Ports ports, Map<String, String> remainingServerOpts)
+            throws IOException, InterruptedException, TimeoutException {
+        return new TypeDBClusterRunner(ports, set(), remainingServerOpts);
+    }
+
+    public static TypeDBClusterRunner create(Ports ports, Set<Ports> peers, Map<String, String> remainingServerOpts)
```
Comment (lolski) [inline test/server/TypeDBClusterRunner.java:74]: Create a runner for a server within a cluster. The server will be started on the configured ports and options, communicating with the configured peers.

File: test/Runner.java [hunk 3]
```diff
@@ -49,7 +49,7 @@ public Runner() throws InterruptedException, TimeoutException, IOException {
             throw new IllegalArgumentException("TypeDB distribution file is missing from " + distributionArchive.getAbsolutePath());
         }
 
-        checkAndDeleteExistingDistribution(distributionArchive);
+        checkAndDeleteExistingDistribution();
```
Comment (lolski) [inline test/Runner.java:52]: Do a lightweight refactor of this class by:
1. Placing the methods which are related close to one another.
2. Removing unneeded method arguments.

File: test/server/TypeDBClusterRunner.java [hunk 4]
```diff
@@ -18,34 +18,70 @@
...
+import static com.vaticle.typedb.common.collection.Collections.set;
 
 public class TypeDBClusterRunner extends TypeDBRunner {
 
-    private final Triple<Integer, Integer, Integer> ports;
-    private final List<Triple<Integer, Integer, Integer>> peerPorts;
+    public static final String OPT_ADDR = "--server.address";
+    private static final String OPT_INTERNAL_ADDR_ZMQ = "--server.internal-address.zeromq";
+    private static final String OPT_INTERNAL_ADDR_GRPC = "--server.internal-address.grpc";
+    public static final String OPT_PEERS_ADDR = "--server.peers.server-{peer-id}.address";
```
Comment (flyingsilverfin) [inline test/server/TypeDBClusterRunner.java:39]: make private

File: test/server/TypeDBClusterRunner.java [hunk 5]
```diff
@@ -18,34 +18,70 @@
...
+    private static final String OPT_PEERS_INTERNAL_ADDR_ZMQ = "--server.peers.server-{peer-id}.internal-address.zeromq";
+    private static final String OPT_PEERS_INTERNAL_ADDR_GRPC = "--server.peers.server-{peer-id}.internal-address.grpc";
```
Comment (flyingsilverfin) [inline test/server/TypeDBClusterRunner.java:41]: could use a string format

File: test/server/TypeDBClusterRunner.java [hunk 6]
```diff
@@ -55,33 +91,100 @@ protected String name() {
...
+            options.put(intAddrZMQKey, host() + ":" + peer.internalZMQ());
+            options.put(intlAddrGRPCKey, host() + ":" + peer.internalGRPC());
+            index++;
+        }
+        return options;
+    }
+
+    public static class Ports {
+        private final int external;
+
```
Comment (flyingsilverfin) [inline test/server/TypeDBClusterRunner.java:155]: no need for newline

File: test/server/TypeDBClusterRunner.java [hunk 7]
```diff
@@ -55,33 +91,100 @@ protected String name() {
...
+        return options;
+    }
+
+    private static Map<String, String> peerOptions(Set<Ports> peers) {
+        Map<String, String> options = new HashMap<>();
+        int index = 0;
+        for (Ports peer: peers) {
+            String addrKey = OPT_PEERS_ADDR.replace("{peer-id}", "peer-" + index);
+            String intAddrZMQKey = OPT_PEERS_INTERNAL_ADDR_ZMQ.replace("{peer-id}", "peer-" + index);
+            String intlAddrGRPCKey = OPT_PEERS_INTERNAL_ADDR_GRPC.replace("{peer-id}", "peer-" + index);
```
Comment (flyingsilverfin) [inline test/server/TypeDBClusterRunner.java:144]: so these can become `OPTS_PEERS_ADDR.format(index)`, where you move `peer-%d` inside of the template

File: test/server/TypeDBClusterRunner.java [hunk 8]
```diff
@@ -55,33 +91,100 @@ protected String name() {
...
 
     @Override
-    public void start() {
-        verifyPortUnused(ports.second());
-        verifyPortUnused(ports.third());
-        super.start();
+    protected void verifyPortUnused() {
+        super.verifyPortUnused(port());
+        super.verifyPortUnused(ports().internalZMQ());
+        super.verifyPortUnused(ports().internalGRPC());
```
Comment (flyingsilverfin) [inline test/server/TypeDBClusterRunner.java:113]: don't need these super calls

File: test/server/TypeDBClusterRunner.java [hunk 9]
```diff
@@ -18,34 +18,70 @@
...
-        this.ports = ports;
-        this.peerPorts = peerPorts;
+    private final Ports ports;
+    private final Set<Ports> peers;
+    private final Map<String, String> remainingServerOpts;
+
+    public static TypeDBClusterRunner create() throws InterruptedException, TimeoutException, IOException {
+        int port = ThreadLocalRandom.current().nextInt(40000, 60000);
+        Ports server = new Ports(port, port+1, port+2);
+        return TypeDBClusterRunner.create(server);
```
Comment (flyingsilverfin) [inline test/server/TypeDBClusterRunner.java:50]: you shouldn't need to call `TypeDBClusterRunner.create(server)`, just call `create(server)` right?

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**